src/exo/cursors.py

```diff
@@ -0,0 +1,228 @@
+from __future__ import annotations
+
+import weakref
+from dataclasses import dataclass
+from enum import Enum
+from typing import Union, List, Iterable
+from weakref import ReferenceType
+
+from .API_types import ProcedureBase
+from .LoopIR import LoopIR
+
+
+class CursorKind(Enum):
+    Node = 0  # Pointing directly at a node
+    # BlockFront = 1  # Gap before the first statement of a block
+    # BlockEnd = 2  # Gap after the last statement of a block
+    GapBefore = 3  # Gap before the node in a block
+    GapAfter  = 4  # Gap after the node in a block
+
+
+class ExoIR(Enum):
+    Assign = 0
+    Reduce = 1
+    WriteConfig = 2
+    Pass = 3
+    If = 4
+    For = 5
+    Alloc = 6
+    Call = 7
+    WindowStmt = 8
+
+
+class InvalidCursorError(Exception):
+    pass
+
+
+@dataclass
+class Cursor:
+    _proc: ReferenceType[ProcedureBase]
+    _node: ReferenceType[Union[LoopIR.proc, LoopIR.stmt, LoopIR.expr]]
+    _path: List[int]
+    _kind: CursorKind = CursorKind.Node
+
+    # ------------------------------------------------------------------------ #
+    # Static constructors
+    # ------------------------------------------------------------------------ #
+
+    @staticmethod
+    def root(proc):
+        return Cursor.from_node(proc, proc.INTERNAL_proc(), [])
+
+    @staticmethod
+    def from_node(proc, node, path):
+        return Cursor(weakref.ref(proc), weakref.ref(node), path)
```

Comment:
This probably ought not be exposed to users, but if `Cursor` objects are exposed to users, then we probably need to expose this class for typechecking.  Therefore a user would be able to call `Cursor.from_node`.

One solution to this problem might be to have a cursor wrapper class in `API.py` and repeat all the user-accessible methods there, wrapping this actual implementation.  Otherwise, we need to indicate that this class is exposed and be more disciplined about underscores on methods.

Reply:
+1 for having a cursor wrapper class in API.py